File: frontend/lib/feedback.ts
/**
 * Types and API call for POST /v1/recommendations/{recommendationId}/feedback
 * (Milestone 7 -- see backend app/schemas/feedback.py / app/api/feedback.py).
 *
 * scoreBreakdown is sent straight from the Recommendation the user is
 * reacting to (see chat.ts) -- the same factor->score map already
 * rendered on the card, not re-derived here. That mirrors the backend
 * schema's own reasoning (feedback.py's module docstring): it's what the
 * user actually saw when they acted, and the backend has nothing durable
 * to re-fetch it from between turns anyway.
 */
import { apiFetch } from "./api-client";

export type FeedbackAction = "accepted" | "rejected";

export interface FeedbackPayload {
  sessionId: string;
  intent: string;
  action: FeedbackAction;
  reason?: string;
  scoreBreakdown: Record<string, number>;
}

export async function sendFeedback(
  idToken: string | null,
  recommendationId: string,
  payload: FeedbackPayload
): Promise<void> {
  const res = await apiFetch(`/v1/recommendations/${encodeURIComponent(recommendationId)}/feedback`, idToken, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });

  if (!res.ok) {
    const detail = await res.json().catch(() => ({}));
    const text = detail.detail
      ? typeof detail.detail === "string"
        ? detail.detail
        : JSON.stringify(detail.detail)
      : `Error ${res.status}`;
    throw new Error(text);
  }
}

File: backend/app/schemas/feedback.py
"""
Feedback request/record schemas (design doc Step 6's
users/{userId}/feedback/{feedbackId} collection, Step 7's
POST /v1/recommendations/{recommendationId}/feedback contract).

score_breakdown is deliberately part of the request, not re-derived
server-side: it's the exact same factor->score map the client already
rendered on the card the user just accepted/rejected (M5.2's
scoreBreakdown). Echoing it back is honest (it's what the user actually
saw when they acted) and cheap (no re-fetch of the original recommendation,
which the backend doesn't keep around after a turn ends anyway -- the
LangGraph checkpointer's state is turn-scoped, not a permanent record of
every recommendation ever shown). M7.3's adjustment rules read this map to
figure out which factors were unfavorable on rejected items, which
favorable on accepted ones -- no fabricated signal, just what was already
computed and shown.
"""
from typing import Literal

from pydantic import Field

from app.schemas.preferences import CamelModel

FeedbackAction = Literal["accepted", "rejected"]


class FeedbackRequest(CamelModel):
    session_id: str
    # Which domain this recommendation came from -- fitness/workspace/route/
    # weather. The adjustment rules (M7.3) only make sense per-domain (e.g.
    # "affordability" isn't a factor route/weather results have at all), so
    # this is required, not inferred later from score_breakdown's keys.
    intent: str
    action: FeedbackAction
    reason: str | None = None
    score_breakdown: dict[str, float] = Field(default_factory=dict)


class FeedbackRecord(CamelModel):
    """What actually gets stored, one layer past the request: adds the
    recommendation/session identifiers the URL path and auth already give
    us, so a stored record is self-contained (readable later without
    needing to know which request produced it)."""

    related_recommendation_id: str
    related_session_id: str
    intent: str
    action: FeedbackAction
    reason: str | None = None
    score_breakdown: dict[str, float] = Field(default_factory=dict)
